core/src/graph/graph.rs

```diff
@@ -563,15 +615,82 @@ impl Graph {
 	/// Determine if page is full
 	///  aggressive:false -> use a simple heuristic based on the number of connections
 	///  aggressive:true  -> do actual compression to determine resulting actual page size
-	pub fn is_page_full(&self, page: &GraphPage, aggressive: bool) -> bool {
+	pub fn try_add_connection_to_page(
+		&self,
+		page: &mut GraphPage,
+		connection_id: &DsnpUserId,
+		aggressive: bool,
+		keys: &Option<ResolvedKeyPair>,
+		dsnp_version_config: Option<&DsnpVersionConfig>,
+	) -> Result<()> {
 		let connection_type = self.get_connection_type();
-		let max_connections_per_page = *PAGE_CAPACITIY_MAP.get(&connection_type).unwrap_or(&50);
+		let max_connections_per_page =
+			*PAGE_CAPACITY_MAP.get(&connection_type).unwrap_or_else(|| {
+				let mut capacities: Vec<&usize> = PAGE_CAPACITY_MAP.values().collect();
+				capacities.sort();
+				capacities.first().unwrap() // default: return smallest capacity value
+			});
 
 		if !aggressive {
-			return page.connections().len() >= max_connections_per_page
+			if page.connections().len() >= max_connections_per_page {
+				return Err(Error::msg("Page trivially full"))
+			}
+
+			if let Err(e) = page.add_connection(connection_id) {
+				return Err(e)
+			}
+
+			return Ok(())
+		}
+
+		if connection_type.privacy_type() == PrivacyType::Private &&
+			(dsnp_version_config.is_none() || keys.is_none())
+		{
+			return Err(Error::msg("Missing encryption key or config"))
+		}
+
+		let max_page_size = self.environment.get_config().max_graph_page_size_bytes as usize;
+		// Max page fullness leaves room for 2 uncompressed connections
```

Comment:
Would you mind add more details where this number 2 comes from?

Reply:
it seems like we would never fill these pages to the max with this logic. the point of having some capacity inside pages was to allow insertions in future if that page got touched by a removal but based on current implementation it seems like this extra capacity will never be filled under any condition

Reply:
Added.